server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java

```diff
@@ -389,8 +390,34 @@ protected void findDefaultServiceOfferingId() {
         serviceOfferingId = serviceOffering.getId();
     }
 
+    protected void findAccountServiceOfferingId(long accountId) {
+        String accountRouterOffering = VirtualNetworkApplianceManager.VirtualRouterServiceOffering.valueIn(accountId);
+        String globalRouterOffering = VirtualNetworkApplianceManager.VirtualRouterServiceOffering.value();
+        if (accountRouterOffering != null) {
+            verifyServiceOfferingByUuid(accountRouterOffering);
+        }
+        if (serviceOfferingId == null && globalRouterOffering != accountRouterOffering) {
```

Comment:
Wouldn't we want `globalRouterOffering` to take precedence over `serviceOfferingId`, @weizhouapache?

Reply:
@DaanHoogland serviceOfferingId is the final offering which will be used on virtual router.

Reply:
yes, but if we configured a global router offering wouldn't we prefer that even if serviceOfferingId is not null?

Reply:
@daanhoogland correct. it checks following offerings

1. Router service Offering of network/vpc offering
2. Account route service offering
3. Global route service offering 
4. Default route offering 

If 1 or 2 is set and validated, global setting will not be used.

Reply:
@weizhouapache `if (serviceOffering != null)`, which should be always, `globalRouterOffering` will not be checked, whether or not `globalRouterOffering` == `accountRouterOffering`.
so I don't think your 3 and 4 are correct but this condition will never be met and `verifyServiceOfferingByUuid(globalRouterOffering)` will never be executed.
So, why check for `serviceOfferingId` before `globalRouterOffering` when it is the last resort and is only relevant after `globalRouterOffering`?

Reply:
> @weizhouapache `if (serviceOffering != null)`, which should be always, `globalRouterOffering` will not be checked, whether or not `globalRouterOffering` == `accountRouterOffering`.
> so I don't think your 3 and 4 are correct but this condition will never be met and `verifyServiceOfferingByUuid(globalRouterOffering)` will never be executed.
> So, why check for `serviceOfferingId` before `globalRouterOffering` when it is the last resort and is only relevant after `globalRouterOffering`?

@DaanHoogland it is serviceOfferingId , not serviceOffering.

https://github.com/apache/cloudstack/blob/3783a092a9431e64f6cf11fc7a1056b338f4f7ab/server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java#L109


the check procedure

https://github.com/apache/cloudstack/blob/3783a092a9431e64f6cf11fc7a1056b338f4f7ab/server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java#L416-L424


in the account part, considering the uuid might be invalid, so check account setting at first, then global setting if account setting is invalid. (if account setting is not set, accountRouterOffering is actually the global setting, so skip the check if accountRouterOffering equals to globalRouterOffering )

https://github.com/apache/cloudstack/blob/3783a092a9431e64f6cf11fc7a1056b338f4f7ab/server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java#L393-L414


hope it helps you understanding all changes in PR.

Reply:
ah, i missed the side effect of `verifyServiceOfferingByUuid()`, where the id is set, thanks.

Maybe to make this more obvious for stupid people like me we can make `verifyServiceOfferingByUuid()` return a uuid or null and set it in the `findAccountServiceOfferingId()`?